RecoEcal/EgammaClusterAlgos/interface/SCEnergyCorrectorSemiParm.h

```diff
@@ -5,80 +5,182 @@
-#include "CondFormats/GBRForest/interface/GBRForestD.h"
-#include "DataFormats/EcalRecHit/interface/EcalRecHit.h"
-#include "DataFormats/EcalRecHit/interface/EcalRecHitCollections.h"
+#include "FWCore/Framework/interface/ESHandle.h"
+#include "FWCore/Framework/interface/ConsumesCollector.h"
+#include "FWCore/ParameterSet/interface/ParameterSet.h"
 #include "Geometry/CaloEventSetup/interface/CaloTopologyRecord.h"
 #include "Geometry/CaloTopology/interface/CaloTopology.h"
 #include "Geometry/Records/interface/CaloGeometryRecord.h"
 #include "Geometry/CaloGeometry/interface/CaloGeometry.h"
-#include "FWCore/Framework/interface/ConsumesCollector.h"
 #include "DataFormats/EgammaReco/interface/SuperCluster.h"
-#include "FWCore/Framework/interface/Event.h"
-#include "FWCore/Framework/interface/ESHandle.h"
-#include "FWCore/ParameterSet/interface/ParameterSet.h"
-#include "FWCore/Utilities/interface/ESGetToken.h"
+#include "DataFormats/EcalRecHit/interface/EcalRecHit.h"
+#include "DataFormats/EcalRecHit/interface/EcalRecHitCollections.h"
+#include "DataFormats/EcalDetId/interface/EcalSubdetector.h"
+#include "DataFormats/ParticleFlowReco/interface/PFRecHit.h"
+#include "DataFormats/VertexReco/interface/VertexFwd.h"
+#include "CondFormats/GBRForest/interface/GBRForestD.h"
 #include "CondFormats/DataRecord/interface/GBRDWrapperRcd.h"
+#include "RecoEgamma/EgammaTools/interface/EgammaBDTOutputTransformer.h"
+#include "RecoEgamma/EgammaTools/interface/HGCalShowerShapeHelper.h"
 
 class SCEnergyCorrectorSemiParm {
 public:
   SCEnergyCorrectorSemiParm();
-  ~SCEnergyCorrectorSemiParm();
+  //if you want override the default on where conditions are consumed, you need to use
+  //the other constructor and then call setTokens approprately
+  SCEnergyCorrectorSemiParm(const edm::ParameterSet& iConfig, edm::ConsumesCollector cc);
+
+  static void fillPSetDescription(edm::ParameterSetDescription& desc);
+  static edm::ParameterSetDescription makePSetDescription();
+
+  template <edm::Transition tr = edm::Transition::BeginLuminosityBlock>
+  void setTokens(const edm::ParameterSet& iConfig, edm::ConsumesCollector cc);
 
-  void setTokens(const edm::ParameterSet &iConfig, edm::ConsumesCollector &cc);
+  void setEventSetup(const edm::EventSetup& es);
+  void setEvent(const edm::Event& e);
 
-  std::pair<double, double> getCorrections(const reco::SuperCluster &sc) const;
-  void modifyObject(reco::SuperCluster &sc);
+  std::pair<double, double> getCorrections(const reco::SuperCluster& sc) const;
+  void modifyObject(reco::SuperCluster& sc) const;
 
-  void setEventSetup(const edm::EventSetup &es);
-  void setEvent(const edm::Event &e);
+  std::vector<float> getRegData(const reco::SuperCluster& sc) const;
 
 protected:
-  const GBRForestD *foresteb_;
-  const GBRForestD *forestee_;
-  const GBRForestD *forestsigmaeb_;
-  const GBRForestD *forestsigmaee_;
+  class RegParam {
+  public:
+    RegParam(std::string meanKey = "",
+             float meanLow = 0,
```

Comment:
```suggestion
    RegParam(std::string meanKey,
             float meanLow,
```
from the use cases I don't see why default values are needed

Reply:
This is the result of two styles clashing. I prefer an initialise at construction approach while SCEnergyCorrectorSemiPara is a construct then initialise paradigm.   

RegParam contains at EgammaBDTOutputTransformer which is has no way of altering its parameters after construction. However the RegParam objects only get their parameters latter. Therefore some defaults must be sent. I could just hard code the defaults into the constructor but to me it makes very little difference and in this way I can eventually adapt the class to my prefered  paradigm.

Note this is the entire reason EgammaBTDOutputTransformer lost the constness of its two member variables, I needed the move constructor active and that was preferable to writing a custom one.

Reply:
I perhaps missed the line(s) in code where RegParam is used without defaults. Please clarify.

Reply:
> I perhaps missed the line(s) in code where RegParam is used without defaults. Please clarify.

sorry, I meant *with defaults*

Reply:
line 39, the default constructor of SCEnergyCorrectorSemiParm. 

So I could get rid of that. But the problem is now we have to declare our transitions for ESProducts which I achieve through templates. But I dont want to template the class, just the method which calls it. 

Now there are other solutions but are a few default parameters really so bad? I'm not seeing the problem.